server/plugin.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/gotenberg"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/topdf"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/x/xhttp"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/x/xtime"
+	"github.com/mattermost/mattermost-server/plugin"
+	"github.com/rs/cors"
+)
+
+// Plugin is a TOPDF app that implements plugin.MattermostPlugin.
+type Plugin struct {
+	plugin.MattermostPlugin
+
+	// app is the actual, underlying TOPDF app and its features exposed to
+	// network via Plugin's HTTP API.
+	app interface {
+		CheckServerStatus() (running bool, err error)
+		GetPDF(userID, fileID string) (pdf io.ReadCloser, err error)
+	} // *topdf.TOPDF
+
+	// gt used by the app to interact with topdf.PDFServer implementation(Gotenberg).
+	// it's also used by Plugin to update Gotenberg's configs when Plugin configs reloaded.
+	gt *gotenberg.Gotenberg
+}
+
+// configuration holds Plugin's config.
+// see plugin.json at root for more info about all configurations.
+type configuration struct {
+	GotenbergAddress        string
+	GotenbergConvertTimeout xtime.Duration
+}
+
+func main() {
+	plugin.ClientMain(&Plugin{
+		gt: gotenberg.New("http://localhost:3000"),
+	})
+}
+
+// OnActivate hook initializes TOPDF app.
+// plugin.MattermostPlugin.API is set just before this hook is called, this is why
+// we're initializing the app at this stage.
+func (p *Plugin) OnActivate() error {
+	p.app = topdf.New(p.MattermostPlugin.API, p.gt)
+	return nil
+}
+
+// OnConfigurationChange hook updates underlying Gotenberg configs.
+func (p *Plugin) OnConfigurationChange() error {
+	var conf configuration
+	if err := p.API.LoadPluginConfiguration(&conf); err != nil {
+		return err
+	}
+	options := []gotenberg.Option{
+		gotenberg.AddrOption(conf.GotenbergAddress),
+		gotenberg.ConvertTimeoutOption(time.Duration(conf.GotenbergConvertTimeout)),
+	}
+	p.gt.UpdateConfig(options...)
```

Comment:
See above, but it seems unfortunate to have to "know" about Gotenberg at this layer -- maybe we just proxy the configuration to the application instead, and let it reconfigure?

Reply:
related to https://github.com/ilgooz/mattermost-plugin-topdf/pull/1#discussion_r346544207 and https://github.com/ilgooz/mattermost-plugin-topdf/pull/1#discussion_r346548396